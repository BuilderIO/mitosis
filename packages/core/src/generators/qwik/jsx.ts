import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
