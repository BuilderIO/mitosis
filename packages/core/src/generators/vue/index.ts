export * from './vue';
export * from './types';
