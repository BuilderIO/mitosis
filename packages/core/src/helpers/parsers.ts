import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
