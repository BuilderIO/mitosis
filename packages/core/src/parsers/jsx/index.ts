export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
