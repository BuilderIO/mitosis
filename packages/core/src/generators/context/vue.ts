import { getContextWithSymbolKey } from './helpers/context-with-symbol-key';

export const contextToVue = getContextWithSymbolKey;
