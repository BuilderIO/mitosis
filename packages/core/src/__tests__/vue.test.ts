import { componentToVue } from '../generators/vue';
import { parseJsx } from '../parsers/jsx';
const basic = require('./data/basic.raw');
const submitButtonBlock = require('./data/blocks/submit-button.raw');
const inputBlock = require('./data/blocks/input.raw');
const selectBlock = require('./data/blocks/select.raw');
const formBlock = require('./data/blocks/form.raw');
const button = require('./data/blocks/button.raw');
const textarea = require('./data/blocks/textarea.raw');
const img = require('./data/blocks/img.raw');
const video = require('./data/blocks/video.raw');
const section = require('./data/blocks/section.raw');
const text = require('./data/blocks/text.raw');
const rawText = require('./data/blocks/raw-text.raw');

describe('Vue', () => {
  test('Basic', () => {
    const json = parseJsx(basic);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Input block', () => {
    const json = parseJsx(inputBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Submit button block', () => {
    const json = parseJsx(submitButtonBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Select block', () => {
    const json = parseJsx(selectBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Form block', () => {
    const json = parseJsx(formBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Button', () => {
    const json = parseJsx(button);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Textarea', () => {
    const json = parseJsx(textarea);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Img', () => {
    const json = parseJsx(img);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Video', () => {
    const json = parseJsx(video);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Section', () => {
    const json = parseJsx(section);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Text', () => {
    const json = parseJsx(text);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('RawText', () => {
    const json = parseJsx(rawText);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });
});
