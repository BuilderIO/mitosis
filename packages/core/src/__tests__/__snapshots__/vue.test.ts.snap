// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue Basic 1`] = `
"<template>
  <div>
    <input :value=\\"name\\" @change=\\"name = $event.target.value\\" />
    Hello! I can run in React, Vue, Solid, or Liquid!
  </div>
</template>
<script>
export default {
  data: () => ({ name: \\"Steve\\" }),
};
</script>
"
`;

exports[`Vue Button 1`] = `
"<template>
  <div>
    <template v-if=\\"link\\">
      <a
        v-bind=\\"attributes\\"
        :href=\\"link\\"
        :target=\\"openLinkInNewTab ? '_blank' : undefined\\"
      >
        {{ text }}
      </a>
    </template>

    <template v-if=\\"!link\\">
      <button v-bind=\\"attributes\\" type=\\"button\\">
        {{ text }}
      </button>
    </template>
  </div>
</template>
<script>
export default {
  props: [\\"link\\", \\"attributes\\", \\"openLinkInNewTab\\", \\"text\\"],
};
</script>
"
`;

exports[`Vue CustomCode 1`] = `
"<template>
  <div
    ref=\\"elem\\"
    :className=\\"
      'builder-custom-code' + (this.replaceNodes ? ' replace-nodes' : '')
    \\"
    :v-html=\\"code\\"
  ></div>
</template>
<script>
export default {
  props: [\\"replaceNodes\\", \\"code\\"],

  data: () => ({ scriptsInserted: [], scriptsRun: [] }),

  mounted() {
    this.findAndRunScripts();
  },

  methods: {
    findAndRunScripts() {
      // TODO: Move this function to standalone one in '@builder.io/utils'
      if (this.$refs.elem && typeof window !== \\"undefined\\") {
        const scripts = this.$refs.elem.getElementsByTagName(\\"script\\");

        for (let i = 0; i < scripts.length; i++) {
          const script = scripts[i];

          if (script.src) {
            if (this.scriptsInserted.includes(script.src)) {
              continue;
            }

            this.scriptsInserted.push(script.src);
            const newScript = document.createElement(\\"script\\");
            newScript.async = true;
            newScript.src = script.src;
            document.head.appendChild(newScript);
          } else if (
            !script.type ||
            [
              \\"text/javascript\\",
              \\"application/javascript\\",
              \\"application/ecmascript\\",
            ].includes(script.type)
          ) {
            if (this.scriptsRun.includes(script.innerText)) {
              continue;
            }

            try {
              this.scriptsRun.push(script.innerText);
              new Function(script.innerText)();
            } catch (error) {
              console.warn(\\"\`CustomCode\`: Error running script:\\", error);
            }
          }
        }
      }
    },
  },
};
</script>
"
`;

exports[`Vue Embed 1`] = `
"<template>
  <div
    ref=\\"elem\\"
    :className=\\"
      'builder-custom-code' + (this.replaceNodes ? ' replace-nodes' : '')
    \\"
    :v-html=\\"code\\"
  ></div>
</template>
<script>
export default {
  props: [\\"replaceNodes\\", \\"code\\"],

  data: () => ({ scriptsInserted: [], scriptsRun: [] }),

  mounted() {
    this.findAndRunScripts();
  },

  methods: {
    findAndRunScripts() {
      // TODO: Move this function to standalone one in '@builder.io/utils'
      if (this.$refs.elem && typeof window !== \\"undefined\\") {
        const scripts = this.$refs.elem.getElementsByTagName(\\"script\\");

        for (let i = 0; i < scripts.length; i++) {
          const script = scripts[i];

          if (script.src) {
            if (this.scriptsInserted.includes(script.src)) {
              continue;
            }

            this.scriptsInserted.push(script.src);
            const newScript = document.createElement(\\"script\\");
            newScript.async = true;
            newScript.src = script.src;
            document.head.appendChild(newScript);
          } else if (
            !script.type ||
            [
              \\"text/javascript\\",
              \\"application/javascript\\",
              \\"application/ecmascript\\",
            ].includes(script.type)
          ) {
            if (this.scriptsRun.includes(script.innerText)) {
              continue;
            }

            try {
              this.scriptsRun.push(script.innerText);
              new Function(script.innerText)();
            } catch (error) {
              console.warn(\\"\`CustomCode\`: Error running script:\\", error);
            }
          }
        }
      }
    },
  },
};
</script>
"
`;

exports[`Vue Form block 1`] = `
"<template>
  <form
    v-bind=\\"attributes\\"
    :validate=\\"validate\\"
    ref=\\"formRef\\"
    :action=\\"!sendWithJs && action\\"
    :method=\\"method\\"
    :name=\\"name\\"
    @submit=\\"onSubmit(event)\\"
  >
    <template v-if=\\"builderBlock && builderBlock.children\\">
      <template v-for=\\"block in builderBlock?.children\\">
        <BuilderBlockComponent
          :key=\\"block.id\\"
          :block=\\"block\\"
        ></BuilderBlockComponent>
      </template>
    </template>

    <template v-if=\\"submissionState === 'error'\\">
      <BuilderBlocks
        dataPath=\\"errorMessage\\"
        :blocks=\\"errorMessage\\"
      ></BuilderBlocks>
    </template>

    <template v-if=\\"submissionState === 'sending'\\">
      <BuilderBlocks
        dataPath=\\"sendingMessage\\"
        :blocks=\\"sendingMessage\\"
      ></BuilderBlocks>
    </template>

    <template v-if=\\"submissionState === 'error' && responseData\\">
      <pre class=\\"builder-form-error-text pre-1\\">
        {{ JSON.stringify(responseData, null, 2) }}
      </pre>
    </template>

    <template v-if=\\"submissionState === 'success'\\">
      <BuilderBlocks
        dataPath=\\"successMessage\\"
        :blocks=\\"successMessage\\"
      ></BuilderBlocks>
    </template>
  </form>
</template>
<script>
import { BuilderBlockComponent as BuilderBlock } from \\"@fake\\";
import { Builder, builder } from \\"@builder.io/sdk\\";
import { BuilderBlocks } from \\"@fake\\";
import { set } from \\"@fake\\";
import { get } from \\"@fake\\";

export default {
  props: [
    \\"previewState\\",
    \\"sendWithJs\\",
    \\"sendSubmissionsTo\\",
    \\"action\\",
    \\"customHeaders\\",
    \\"contentType\\",
    \\"sendSubmissionsToEmail\\",
    \\"name\\",
    \\"method\\",
    \\"errorMessagePath\\",
    \\"resetFormOnSubmit\\",
    \\"successUrl\\",
    \\"validate\\",
    \\"attributes\\",
    \\"builderBlock\\",
    \\"errorMessage\\",
    \\"sendingMessage\\",
    \\"successMessage\\",
  ],

  data: () => ({
    state: \\"unsubmitted\\",
    responseData: null,
    formErrorMessage: \\"\\",
    BuilderBlock,
    Builder,
    builder,
    BuilderBlocks,
    set,
    get,
  }),

  computed: {
    submissionState() {
      return (Builder.isEditing && this.previewState) || this.state;
    },
  },

  methods: {
    onSubmit(event) {
      const sendWithJs = this.sendWithJs || this.sendSubmissionsTo === \\"email\\";

      if (this.sendSubmissionsTo === \\"zapier\\") {
        event.preventDefault();
      } else if (sendWithJs) {
        if (!(this.action || this.sendSubmissionsTo === \\"email\\")) {
          event.preventDefault();
          return;
        }

        event.preventDefault();
        const el = event.currentTarget;
        const headers = this.customHeaders || {};
        let body;
        const formData = new FormData(el); // TODO: maybe support null

        const formPairs = Array.from(
          event.currentTarget.querySelectorAll(\\"input,select,textarea\\")
        )
          .filter((el) => !!el.name)
          .map((el) => {
            let value;
            const key = el.name;

            if (el instanceof HTMLInputElement) {
              if (el.type === \\"radio\\") {
                if (el.checked) {
                  value = el.name;
                  return {
                    key,
                    value,
                  };
                }
              } else if (el.type === \\"checkbox\\") {
                value = el.checked;
              } else if (el.type === \\"number\\" || el.type === \\"range\\") {
                const num = el.valueAsNumber;

                if (!isNaN(num)) {
                  value = num;
                }
              } else if (el.type === \\"file\\") {
                // TODO: one vs multiple files
                value = el.files;
              } else {
                value = el.value;
              }
            } else {
              value = el.value;
            }

            return {
              key,
              value,
            };
          });
        let contentType = this.contentType;

        if (this.sendSubmissionsTo === \\"email\\") {
          contentType = \\"multipart/form-data\\";
        }

        Array.from(formPairs).forEach(({ value }) => {
          if (
            value instanceof File ||
            (Array.isArray(value) && value[0] instanceof File) ||
            value instanceof FileList
          ) {
            contentType = \\"multipart/form-data\\";
          }
        }); // TODO: send as urlEncoded or multipart by default
        // because of ease of use and reliability in browser API
        // for encoding the form?

        if (contentType !== \\"application/json\\") {
          body = formData;
        } else {
          // Json
          const json = {};
          Array.from(formPairs).forEach(({ value, key }) => {
            set(json, key, value);
          });
          body = JSON.stringify(json);
        }

        if (contentType && contentType !== \\"multipart/form-data\\") {
          if (
            // Zapier doesn't allow content-type header to be sent from browsers
            !(sendWithJs && this.action?.includes(\\"zapier.com\\"))
          ) {
            headers[\\"content-type\\"] = contentType;
          }
        }

        const presubmitEvent = new CustomEvent(\\"presubmit\\", {
          detail: {
            body,
          },
        });

        if (this.$refs.formRef) {
          this.$refs.formRef.dispatchEvent(presubmitEvent);

          if (presubmitEvent.defaultPrevented) {
            return;
          }
        }

        this.state = \\"sending\\";
        const formUrl = \`\${
          builder.env === \\"dev\\" ? \\"http://localhost:5000\\" : \\"https://builder.io\\"
        }/api/v1/form-submit?apiKey=\${builder.apiKey}&to=\${btoa(
          this.sendSubmissionsToEmail || \\"\\"
        )}&name=\${encodeURIComponent(this.name || \\"\\")}\`;
        fetch(
          this.sendSubmissionsTo === \\"email\\" ? formUrl : this.action, // TODO: throw error if no action URL
          {
            body,
            headers,
            method: this.method || \\"post\\",
          }
        ).then(
          async (res) => {
            let body;
            const contentType = res.headers.get(\\"content-type\\");

            if (contentType && contentType.indexOf(\\"application/json\\") !== -1) {
              body = await res.json();
            } else {
              body = await res.text();
            }

            if (!res.ok && this.errorMessagePath) {
              // TODO: allow supplying an error formatter function
              let message = get(body, this.errorMessagePath);

              if (message) {
                if (typeof message !== \\"string\\") {
                  // TODO: ideally convert json to yaml so it woul dbe like
                  // error: - email has been taken
                  message = JSON.stringify(message);
                }

                this.formErrorMessage = message;
              }
            }

            this.responseData = body;
            this.state = res.ok ? \\"success\\" : \\"error\\";

            if (res.ok) {
              const submitSuccessEvent = new CustomEvent(\\"submit:success\\", {
                detail: {
                  res,
                  body,
                },
              });

              if (this.$refs.formRef) {
                this.$refs.formRef.dispatchEvent(submitSuccessEvent);

                if (submitSuccessEvent.defaultPrevented) {
                  return;
                } // TODO: option to turn this on/off?

                if (this.resetFormOnSubmit !== false) {
                  this.$refs.formRef.reset();
                }
              } // TODO: client side route event first that can be preventDefaulted

              if (this.successUrl) {
                if (this.$refs.formRef) {
                  const event = new CustomEvent(\\"route\\", {
                    detail: {
                      url: this.successUrl,
                    },
                  });
                  this.$refs.formRef.dispatchEvent(event);

                  if (!event.defaultPrevented) {
                    location.href = this.successUrl;
                  }
                } else {
                  location.href = this.successUrl;
                }
              }
            }
          },
          (err) => {
            const submitErrorEvent = new CustomEvent(\\"submit:error\\", {
              detail: {
                error: err,
              },
            });

            if (this.$refs.formRef) {
              this.$refs.formRef.dispatchEvent(submitErrorEvent);

              if (submitErrorEvent.defaultPrevented) {
                return;
              }
            }

            this.responseData = err;
            this.state = \\"error\\";
          }
        );
      }
    },
  },
};
</script>
<style>
.pre-1 {
  padding: 10px;
  color: red;
  text-align: center;
}
</style>
"
`;

exports[`Vue Img 1`] = `
"<template>
  <img
    v-bind=\\"attributes\\"
    :style=\\"{
      objectFit: backgroundSize || 'cover',
      objectPosition: backgroundPosition || 'center',
    }\\"
    :key=\\"(Builder.isEditing && imgSrc) || 'default-key'\\"
    :alt=\\"altText\\"
    :src=\\"imgSrc\\"
  />
</template>
<script>
import { Builder } from \\"@builder.io/sdk\\";

export default {
  props: [
    \\"backgroundSize\\",
    \\"backgroundPosition\\",
    \\"attributes\\",
    \\"imgSrc\\",
    \\"altText\\",
  ],

  data: () => ({ Builder }),
};
</script>
"
`;

exports[`Vue Input block 1`] = `
"<template>
  <input
    v-bind=\\"attributes\\"
    :key=\\"Builder.isEditing && defaultValue ? defaultValue : 'default-key'\\"
    :placeholder=\\"placeholder\\"
    :type=\\"type\\"
    :name=\\"name\\"
    :value=\\"value\\"
    :defaultValue=\\"defaultValue\\"
    :required=\\"required\\"
  />
</template>
<script>
import { Builder } from \\"@builder.io/sdk\\";

export default {
  props: [
    \\"attributes\\",
    \\"defaultValue\\",
    \\"placeholder\\",
    \\"type\\",
    \\"name\\",
    \\"value\\",
    \\"required\\",
  ],

  data: () => ({ Builder }),
};
</script>
"
`;

exports[`Vue RawText 1`] = `
"<template>
  <span
    :class=\\"attributes?.class || attributes?.className\\"
    :v-html=\\"text || ''\\"
  ></span>
</template>
<script>
export default {
  props: [\\"attributes\\", \\"text\\"],
};
</script>
"
`;

exports[`Vue Section 1`] = `
"<template>
  <section
    v-bind=\\"attributes\\"
    :style=\\"
      maxWidth && typeof maxWidth === 'number'
        ? {
            maxWidth: maxWidth,
          }
        : undefined
    \\"
  >
    <slot></slot>
  </section>
</template>
<script>
export default {
  props: [\\"attributes\\", \\"maxWidth\\"],
};
</script>
"
`;

exports[`Vue Select block 1`] = `
"<template>
  <select
    v-bind=\\"attributes\\"
    :value=\\"value\\"
    :key=\\"Builder.isEditing && defaultValue ? defaultValue : 'default-key'\\"
    :defaultValue=\\"defaultValue\\"
    :name=\\"name\\"
  >
    <template v-for=\\"option in options\\">
      <option :value=\\"option.value\\">{{ option.name || option.value }}</option>
    </template>
  </select>
</template>
<script>
import { Builder } from \\"@builder.io/sdk\\";

export default {
  props: [\\"attributes\\", \\"value\\", \\"defaultValue\\", \\"name\\", \\"options\\"],

  data: () => ({ Builder }),
};
</script>
"
`;

exports[`Vue Stamped.io 1`] = `
"<template>
  <div>
    <button @click=\\"showReviewPrompt = true\\">Write a review</button>
    <template v-if=\\"showReviewPrompt\\">
      <input placeholder=\\"Email\\" />

      <input placeholder=\\"Title\\" class=\\"input-1\\" />

      <textarea
        placeholder=\\"How was your experience?\\"
        class=\\"textarea-1\\"
      ></textarea>

      <button class=\\"button-1\\" @click=\\"showReviewPrompt = false\\">Submit</button>
    </template>
    <template v-for=\\"review in reviews\\">
      <div class=\\"div-1\\">
        <img class=\\"img-1\\" :src=\\"review.avatar\\" />
        <div>
          <div>{{ review.author }}</div>
          <div>{{ review.reviewMessage }}</div>
        </div>
      </div>
    </template>
  </div>
</template>
<script>
export default {
  props: [\\"apiKey\\", \\"productId\\"],

  data: () => ({ reviews: [], showReviewPrompt: false }),

  mounted() {
    fetch(
      \`https://stamped.io/api/widget/reviews?storeUrl=builder-io.myshopify.com&apiKey=\${
        this.apiKey || \\"pubkey-8bbDq7W6w4sB3OWeM1HUy2s47702hM\\"
      }&productId=\${this.productId || \\"2410511106127\\"}\`
    )
      .then((res) => res.json())
      .then((data) => {
        this.reviews = data.data;
      });
  },
};
</script>
<style>
.input-1 {
  display: block;
}
.textarea-1 {
  display: block;
}
.button-1 {
  display: block;
}
.div-1 {
  margin: 10px;
  padding: 10px;
  background: white;
  display: flex;
  border-radius: 5px;
  box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
}
.img-1 {
  height: 30px;
  width: 30px;
  margin-right: 10px;
}
</style>
"
`;

exports[`Vue Submit button block 1`] = `
"<template>
  <button v-bind=\\"attributes\\" type=\\"submit\\">
    {{ text }}
  </button>
</template>
<script>
export default {
  props: [\\"attributes\\", \\"text\\"],
};
</script>
"
`;

exports[`Vue Text 1`] = `
"<template>
  <div
    :contentEditable=\\"allowEditingText || undefined\\"
    :v-html=\\"text || content || ''\\"
  ></div>
</template>
<script>
import { Builder } from \\"@builder.io/sdk\\";

export default {
  props: [\\"text\\", \\"content\\"],

  data: () => ({ Builder }),
};
</script>
"
`;

exports[`Vue Textarea 1`] = `
"<template>
  <textarea
    v-bind=\\"attributes\\"
    :placeholder=\\"placeholder\\"
    :name=\\"name\\"
    :value=\\"value\\"
    :defaultValue=\\"defaultValue\\"
  ></textarea>
</template>
<script>
export default {
  props: [\\"attributes\\", \\"placeholder\\", \\"name\\", \\"value\\", \\"defaultValue\\"],
};
</script>
"
`;

exports[`Vue Video 1`] = `
"<template>
  <video
    v-bind=\\"attributes\\"
    preload=\\"none\\"
    :style=\\"{
      width: '100%',
      height: '100%',
      ...attributes?.style,
      objectFit: fit,
      objectPosition: position,
      // Hack to get object fit to work as expected and
      // not have the video overflow
      borderRadius: 1,
    }\\"
    :key=\\"video || 'no-src'\\"
    :poster=\\"posterImage\\"
    :autoPlay=\\"autoPlay\\"
    :muted=\\"muted\\"
    :controls=\\"controls\\"
    :loop=\\"loop\\"
  ></video>
</template>
<script>
export default {
  props: [
    \\"attributes\\",
    \\"fit\\",
    \\"position\\",
    \\"video\\",
    \\"posterImage\\",
    \\"autoPlay\\",
    \\"muted\\",
    \\"controls\\",
    \\"loop\\",
  ],
};
</script>
"
`;
